lib/streamlit/runtime/fragment.py

```diff
@@ -164,8 +157,18 @@ def wrapped_fragment():
             ctx = get_script_run_ctx(suppress_warning=True)
             assert ctx is not None
 
-            dg_stack.set(deepcopy(dg_stack_snapshot))
-            ctx.current_fragment_id = fragment_id
+            if ctx.current_fragment_id:
```

Comment:
Under what scenario can the fragment id already be set here? If a fragment is used inside another fragment?

Reply:
We now set `ctx.current_fragment_id` when calling `ctx.reset` at the beginning of a script run. The reason this is useful is that we want to be able to differentiate between running a full script or just a fragment since that determines whether we want to restore `ctx.cursors`/`dg_stack` from the snapshot of the two that we saved when this fragment was first declared.

The comments below describe this a bit but could probably be improved on.